### PR TITLE
430 enhancement speed up hantushstep

### PR DESCRIPTION
**Pull request by dbrakenhoff**
# Short Description
Speed up of `Hantush*.step()` and log transform of parameter b for HantushWellModel.
Add option to calculate Hantush using numerical integration (`quad=True`). This implementation yields more accurate parameter estimates when comparing to TTim models. It is much slower though. 

Details:
- add `quad` and `use_numba` options to Hantush response functions
- add numpy_step(), numba_step() and quad_step() methods
- add _hantush integrand for quad() implementation
- modify step() to select calculation method according to options
- log transform parameter b for HantushWellModel (because of small values of b)
- update hantush notebook with options and figure comparing results
- update response function nb for log transformed b in HantushWellModel

# Checklist before PR can be merged:
- [x] closes issue #430
- [x] is documented
- [x] PEP8 compliant code
- [x] tests added / passed
- [x] Example Notebook (for new features)
- [x] API changes documented in Release Notes

### CODE REVIEW

**Comment by dbrakenhoff** [comment]
There needs to be a method to deal with older pas-files as the parameter defintion of $b$ in HantushWellModel has changed. This should be adjusted on load and print an info (or warning) message.

**Comment by mbakker7** [comment]
There are several other options to explore how to make the Hantush function
more accurate but also fast enough. Should we do that first or finish this
pull request and do that next?

**Comment by dbrakenhoff** [comment]
@mbakker7 probably smart to look into that now. The implementation in this PR serves as a good benchmark for comparison purposes.

**Comment by dbrakenhoff** [comment]
Agreed with @mbakker7 that we can implement these changes to Hantush. Testing the IVP approach did not provide a significant improvement over numpy/numba (it was fast but less accurate for small times).

An overview of changes and how they work:

Hantush and HantushWellModel options:
- the `@njit` decorator now optionally takes optional arguments (e.g. `@njit` or `@njit(parallel=True)`)
- add `quad` option, this aproach is more accurate (compared to e.g. TTim) but much slower.
   - add `_hantush_integrand ` for `quad_step`() implementation
- add `use_numba` options to Hantush response functions. `use_numba` requires `numba-scipy` to be installed. Users won't be faced with any errors if `use_numba=True` but `numba-scipy` is not installed, but will see a warning that Hantush is falling back on the numpy implementation. 
- add `numpy_step()`, `numba_step()` and `quad_step()` methods to make the switching detailed above possible
- modify `step()` to select calculation method according to options.

HantushWellModel parameters:
- log transform parameter b for HantushWellModel (because of small values of b): `b_new = np.log(b_old)`. This was shown to help speed up optimizing models in some cases for a project I was working on.
- For older stored HantushWellModels in pas files, the stored parameter `b` is automatically updated on load. A warning is shown that this has happened, similar to the StressModel2 -> RechargeModel warning. Test is added to check this.
- Modify `variance_gain()` and `gain()` formulas accordingly.

Notebooks
- update Hantush notebook with options and figure comparing results
- update response function notebook for log transformed b in HantushWellModel
- update multiple wells notebook with extra comparison plot

Other small fixes
- remove legend from response functions subplots in CompareModels plot (took up too much space and redundant)
- fix issue where heteroscedasticity plots did not work because simulation was not interpolated to oseries index

**Comment by dbrakenhoff** [comment]
Any chance we can merge this soon :innocent:  ?

**Comment by mbakker7** [comment]
Hadn’t we decided to leave it as far as it is now? Alternatives did not
provide a clear advantage.

**Comment by raoulcollenteur** [comment]
I agree with @dbrakenhoff that we can merge this PR. The PR does not make any changes to Hantush, but only provides additional options. Only for HantushWellModel the parameter b changes. 

@mbakker7, we can work on the the alternatives in the future.

@dbrakenhoff, there are a couple of conflicts with the Dev-branch, could you fix those?

Cheers,
Raoul

**Comment by martinvonk** [comment]
Quickly fixed the merge conflicts but now the same problem arises with the docstring build as #441. 

> Exception occurred:
  File "/home/docs/checkouts/readthedocs.org/user_builds/pastas/envs/433/lib/python3.10/site-packages/sphinx/util/nodes.py", line 271, in get_node_source
    raise ValueError("node source not found")
ValueError: node source not found

**Comment by martinvonk** [comment]
All problems with readthedocs are resolved so this can be reviewed and merged.

File: doc/examples/10_multiple_wells.ipynb [hunk 1]
```diff
@@ -45,7 +45,7 @@
     "xo = 85850\n",
     "yo = 383362\n",
     "\n",
-    "# Specify coördinates extractions\n",
```
Comment (raoulcollenteur) [inline doc/examples/10_multiple_wells.ipynb:48]: my favourite change of the PR :)

File: pastas/io/base.py [hunk 2]
```diff
@@ -100,6 +102,19 @@ def _load_model(data):
...
+             version.parse("0.22.0"))):
+            logger.warning("The value of parameter 'b' in HantushWellModel"
+                           "was modified in 0.22.0: b_new = log(b_old). The value of "
+                           "'b' is automatically updated on load.")
+            wnam = ts["name"]
+            for pcol in ["initial", "optimal", "pmin", "pmax"]:
+                if data["parameters"].loc[wnam + "_b", pcol] > 0:
+                    data["parameters"].loc[wnam + "_b", pcol] = \
+                        log(data["parameters"].loc[wnam + "_b", pcol])
+
```
Comment (raoulcollenteur) [inline pastas/io/base.py:117]: This code should be removed at pastas 1.0 I think. just for future reference. Issue #421

**Review by raoulcollenteur — APPROVED** [review]

**Comment by raoulcollenteur** [comment]
Thanks for the PR and all the work on this. Happy to merge this now, as it contains many useful changes and waiting much longer will make it harder to merge. Merging :)

Cheers,
Raoul

**Pull request merged**